index.bs

```diff
@@ -679,11 +681,117 @@ are traditionally added to {{WindowEventHandlers}} rather than {{Window}}.</p>
+If the algorithmic complexity and call graph is reasonably understood, add appropriate
+state to track that the algorithm is in progress and terminate immediately if the algorithm
+is already concurrently running. This technique is "guarding" the algorithm:
+
+<div class="example">
+    A technique (e.g., as used in the [=AbortSignal/add|AbortSignal's add algorithm=]
+    or as a stack-based variation in the [=focus update steps=])
+    for guarding against unplanned recursion. 
+    
+    In this example, an object of type <dfn noexport interface>MyObject</dfn> has
+    a method <dfn method for="MyObject">doComplexOpWithEvents()</dfn> and
+    an internal state attribute [[<dfn noexport attribute for="MyObject">started</dfn>]]
+    referenced below as
+    {{MyObject|this}}.[[{{started}}]] whose initial
+    value is `false`.
+
+    The {{doComplexOpWithEvents()}} method must act as follows:
+    
+    1. If {{MyObject|this}}.[[{{started}}]] is not `false` then throw an {{InvalidStateError}}
+        {{DOMException}} and terminate this algorithm.
+        
+        Note: {{doComplexOpWithEvents()}} is not allowed to be run again
+        (e.g., from within a `"currentlyinprogress"` event handler) before it finishes
+        its original execution.
+
+    2. Set {{MyObject|this}}.[[{{started}}]] to `true`.
+
+    3. ...Do complex stuff...
+
+    4. [=Fire an event=] named `"currentlyinprogress"` at {{MyObject|this}}.
+
+    5. ...Finish up complex stuff...
+
+    6. Set {{MyObject|this}}.[[{{started}}]] to `false`.
+</div>
+
+Sometimes "guarding" as illustrated above cannot be done because of the complexity
+of the algorithm, the number of potential algorithm entry-points, or the side-effects
+beyond the algorithm's control. In these cases another form of protecting the integrity
+of the algorithm's state is to "defer" the event to a subsequent task or microtask.
+Deferral ensures that any stack-based recursion is avoided (but does not eliminate
+potentially problematic loops, as they could now occur as unending follow-up tasks).
+
+Defering an event is often specified as "[=queue a task=] to [=fire an event=]...".
+
+By way of illustration, the deprecated
+[Mutation Events](https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Mutation_events),
+were specified to fire in the middle of node insertion and removal algorithms, which
+are triggered by a variety of common API calls like {{Node/appendChild()}}. The side-effects
+from recursivly running node mutation algorithms caused many years of security issues.
+A more robust approach is embodied in
+{{MutationObserver}}
+applies the principle of deferral of notification — in this case to the microtask queue
+following the algorithm.
+
+Weigh the trade-offs when considering deferral as an alternative to guarding as noted
+earlier. Guarding is generally the preferred approach. With guarding:
```

Comment:
I'm a little nervous about giving advice this strong about preferring guarding.  In particular, I think having an algorithm unexpectedly throw a "you can't do this now" exception can be problematic (both in terms of handling the exception, and in terms of the algorithm that needed to be run not actually being run).  It seems like some of the cases where such exceptions would be fired would be caught quickly in testing... but on the other hand I suspect some of them wouldn't be.  Perhaps it's at least worth mentioning this downside a little bit more?

Reply:
Can you clarify a bit more how this concern is different from any algorithm that wants to unexpectedly throw an exception? I agree that running afoul of the guard can be unexpected, but (at least in this section) the guard is there to prevent unwanted recursion into the algorithm, which usually means you'll still be executing the algorithm, just not the recursive part. (Also IIRC, in some cases the guard will simply terminate an algorithm without throwing an exception.)

Reply:
The concern about bailing out generally (whether an exception or not):  If the algorithm is one that's needed to ensure some aspect of the application's state is up-to-date, failing to run the algorithm could lead to that state being out-of-date for an extended period of time, leading to user-visible bugs (possibly only in edge cases).  (Then again, there are similar sequences of evets that lead to user-visible bugs only in edge cases with the deferral approach.)

Specifically for exceptions: there are many calls where developers don't need to consider the possibility of exceptions.  There are some that throw exceptions rarely where one would reasonably expect errors to occur some of the time (e.g., network failures), but it seems less then ideal for things to fail a low portion of the time in ways that people wouldn't expect.

Reply:
I think this is a fair caution to express, lest guarding be seen as almost always the right solution. I'll try to condense it down.

Reply:
I added two cautionary notes below the example algorithm that I think touch on these concerns. I also softened the language about recommending guarding to a more neutral "there are tradeoffs to both approaches" which leaves the reader with a little less prescriptive guidance and more opportunity to think about which approach works given their situation.